File: trabalho-de-so/banqueiro.py
class BankersAlgorithm:
    def __init__(self, num_processes, num_resources, max_matrix, allocation_matrix, available_resources):
        self.num_processes = num_processes
        self.num_resources = num_resources
        self.max_matrix = max_matrix
        self.allocation_matrix = allocation_matrix
        self.available_resources = available_resources
        
        self.need_matrix = [
            [self.max_matrix[i][j] - self.allocation_matrix[i][j] 
             for j in range(num_resources)] 
            for i in range(num_processes)
        ]
    
    def is_safe_state(self):
        work = self.available_resources.copy()
        finish = [False] * self.num_processes
        safe_sequence = []
        
        for _ in range(self.num_processes):
            found = False
            
            for p in range(self.num_processes):
                if not finish[p] and all(self.need_matrix[p][j] <= work[j] for j in range(self.num_resources)):
                    for j in range(self.num_resources):
                        work[j] += self.allocation_matrix[p][j]
                    
                    finish[p] = True
                    safe_sequence.append(p)
                    found = True
                    break
            
            if not found:
                return False, []
        
        return True, safe_sequence
    
    def request_resources(self, process_id, request):
        if any(request[j] > self.need_matrix[process_id][j] for j in range(self.num_resources)):
            raise ValueError("Request exceeds maximum need")
        
        if any(request[j] > self.available_resources[j] for j in range(self.num_resources)):
            return False
        
        for j in range(self.num_resources):
            self.available_resources[j] -= request[j]
            self.allocation_matrix[process_id][j] += request[j]
            self.need_matrix[process_id][j] -= request[j]
        
        is_safe, _ = self.is_safe_state()
        
        if not is_safe:
            for j in range(self.num_resources):
                self.available_resources[j] += request[j]
                self.allocation_matrix[process_id][j] -= request[j]
                self.need_matrix[process_id][j] += request[j]
            return False
        
        return True

def main():
    print("Test Case 1: Safe State Scenario")
    num_processes = 5
    num_resources = 3
    
    max_matrix = [
        [7, 5, 3],
        [3, 2, 2],
        [9, 0, 2],
        [2, 2, 2],
        [4, 3, 3]
    ]
    
    allocation_matrix = [
        [0, 1, 0],
        [2, 0, 0],
        [3, 0, 2],
        [2, 1, 1],
        [0, 0, 2]
    ]
    
    available_resources = [3, 3, 2]
    
    banker = BankersAlgorithm(num_processes, num_resources, max_matrix, allocation_matrix, available_resources)
    
    is_safe, safe_sequence = banker.is_safe_state()
    print(f"Is Safe State: {is_safe}")
    print(f"Safe Sequence: {safe_sequence}")
    
    test_request = [1, 0, 2]
    print("\nTesting Resource Request:")
    request_result = banker.request_resources(0, test_request)
    print(f"Request Granted: {request_result}")
    
    print("\nTest Case 2: Unsafe State Scenario")
    max_matrix_unsafe = [
        [6, 4, 3],
        [3, 2, 2],
        [9, 0, 2],
        [2, 2, 2],
        [4, 3, 3]
    ]
    
    allocation_matrix_unsafe = [
        [0, 1, 0],
        [3, 0, 2],
        [3, 0, 2],
        [2, 1, 1],
        [0, 0, 2]
    ]
    
    available_resources_unsafe = [1, 1, 0]
    
    banker_unsafe = BankersAlgorithm(num_processes, num_resources, 
                                     max_matrix_unsafe, 
                                     allocation_matrix_unsafe, 
                                     available_resources_unsafe)
    
    is_safe_unsafe, safe_sequence_unsafe = banker_unsafe.is_safe_state()
    print(f"Is Safe State: {is_safe_unsafe}")
    print(f"Safe Sequence: {safe_sequence_unsafe}")

if __name__ == "__main__":
    main()

File: trabalho-de-so/banqueiro.ts
class BankersAlgorithm {
    numProcesses: number;
    numResources: number;
    maxMatrix: number[][];
    allocationMatrix: number[][];
    availableResources: number[];
    needMatrix: number[][];

    constructor(numProcesses: number, numResources: number, maxMatrix: number[][], allocationMatrix: number[][], availableResources: number[]) {
        this.numProcesses = numProcesses;
        this.numResources = numResources;
        this.maxMatrix = maxMatrix;
        this.allocationMatrix = allocationMatrix;
        this.availableResources = availableResources;

        this.needMatrix = this.maxMatrix.map((maxRow, i) =>
            maxRow.map((maxVal, j) => maxVal - this.allocationMatrix[i][j])
        );
    }
    
    isSafeState(): [boolean, number[]] {
        const work = [...this.availableResources];
        const finish = Array(this.numProcesses).fill(false);
        const safeSequence: number[] = [];
    
        for (let i = 0; i < this.numProcesses; i++) {
            let found = false;
    
            for (let p = 0; p < this.numProcesses; p++) {
                let canExecute = true;

                for (let j = 0; j < this.numResources; j++) {
                    if (this.needMatrix[p][j] > work[j]) {
                        canExecute = false;
                        break;
                    }
                }
    
                if (!finish[p] && canExecute) {
                    for (let j = 0; j < this.numResources; j++) {
                        work[j] += this.allocationMatrix[p][j];
                    }
    
                    finish[p] = true;
                    safeSequence.push(p);
                    found = true;
                    break;
                }
            }
    
            if (!found) {
                return [false, []];
            }
        }
    
        return [true, safeSequence];
    }

    requestResources(processId: number, request: number[]): boolean {
        if (request.some((req, j) => req > this.needMatrix[processId][j])) {
            throw new Error("Request exceeds maximum need");
        }

        if (request.some((req, j) => req > this.availableResources[j])) {
            return false;
        }

        for (let j = 0; j < this.numResources; j++) {
            this.availableResources[j] -= request[j];
            this.allocationMatrix[processId][j] += request[j];
            this.needMatrix[processId][j] -= request[j];
        }


        const [isSafe] = this.isSafeState();

        if (!isSafe) {
            for (let j = 0; j < this.numResources; j++) {
                this.availableResources[j] += request[j];
                this.allocationMatrix[processId][j] -= request[j];
                this.needMatrix[processId][j] += request[j];
            }
            return false;
        }

        return true;
    }
}


function main() {
    console.log("Test Case 1: Safe State Scenario");
    const numProcesses = 5;
    const numResources = 3;

    const maxMatrix: number[][] = [
        [7, 5, 3],
        [3, 2, 2],
        [9, 0, 2],
        [2, 2, 2],
        [4, 3, 3]
    ];

    const allocationMatrix: number[][] = [
        [0, 1, 0],
        [2, 0, 0],
        [3, 0, 2],
        [2, 1, 1],
        [0, 0, 2]
    ];

    const availableResources: number[] = [3, 3, 2];

    const banker = new BankersAlgorithm(numProcesses, numResources, maxMatrix, allocationMatrix, availableResources);

    const [isSafe, safeSequence] = banker.isSafeState();
    console.log(`Is Safe State: ${isSafe}`);
    console.log(`Safe Sequence: ${safeSequence}`);

    const testRequest: number[] = [1, 0, 2];
    console.log("\nTesting Resource Request:");
    const requestResult = banker.requestResources(0, testRequest);
    console.log(`Request Granted: ${requestResult}`);

    console.log("\nTest Case 2: Unsafe State Scenario");
    const maxMatrixUnsafe: number[][] = [
        [6, 4, 3],
        [3, 2, 2],
        [9, 0, 2],
        [2, 2, 2],
        [4, 3, 3]
    ];

    const allocationMatrixUnsafe: number[][] = [
        [0, 1, 0],
        [3, 0, 2],
        [3, 0, 2],
        [2, 1, 1],
        [0, 0, 2]
    ];

    const availableResourcesUnsafe: number[] = [1, 1, 0];

    const bankerUnsafe = new BankersAlgorithm(numProcesses, numResources, maxMatrixUnsafe, allocationMatrixUnsafe, availableResourcesUnsafe);

    const [isSafeUnsafe, safeSequenceUnsafe] = bankerUnsafe.isSafeState();
    console.log(`Is Safe State: ${isSafeUnsafe}`);
    console.log(`Safe Sequence: ${safeSequenceUnsafe}`);
}

main();
